[xmppclient] Binding v5.1.3-5.2.0 not working

[Full problem review](https://community.openhab.org/t/xmppclient-binding-v5-1-3-5-2-0-not-working/168974)

Version 5.1.3 is completely broken due to dependency issues.

Version 5.2.0 is completely/partially broken. It can’t start (99% in a Waiting status), and it can’t send messages (publishGroupXMPP) or images (publishXMPPImageByHTTP). I checked the code of 5.0.2-M2.

publishGroupXMPP can be easy fixed .

XMPPClientHandlerFactory.java
```java
@Activate
    protected void activate(ComponentContext componentContext) {
        super.activate(componentContext);
        Roster.setRosterLoadedAtLoginDefault(false);

        DNSUtil.setDNSResolver(JavaxResolver.getInstance());
        SmackConfiguration.setDefaultHostnameVerifier(new XmppHostnameVerifier());
        Base64.setEncoder(Java7Base64Encoder.getInstance());
        Base64UrlSafeEncoder.setEncoder(Java7Base64UrlSafeEncoder.getInstance());

        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/disco#info", new DiscoverInfoProvider());
        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/disco#items", new DiscoverItemsProvider());
        ProviderManager.addExtensionProvider("x", "jabber:x:data", new DataFormProvider());

        // Init MUC
        ProviderManager.addExtensionProvider("x", "http://jabber.org/protocol/muc#user",
                new org.jivesoftware.smackx.muc.provider.MUCUserProvider());
        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/muc#admin",
                new org.jivesoftware.smackx.muc.provider.MUCAdminProvider());
        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/muc#owner",
                new org.jivesoftware.smackx.muc.provider.MUCOwnerProvider());
    }
```

publishXMPPImageByHTTP - I couldn’t fix it, I don’t have enough knowledge

I’m currently using version 5.0.3 build on smak 4.3.3, as 4.5.0-rc1 has some issues with osgi (as it seemed to me).
My own build (5.0.3) work with 5.0.3,5.1.3,5.2.0 just fine. Added MUC join/read support. I want to send a pull request when everything is fixed in the base version.

> Thanks, just added a commit, could you test that again. There discovery is now forced. It also adds more logging to see details on how far it got.
> 
> Are you sure the server that you are connected to supports file uploads ?

Yes. This is my own local OpenFire v5.0.4 server. We use it with our whole family/relatives/friends every day.
Xmppclient 5.0.3 sends files just fine. 
Windows (Gajim) and Android (Conversation) clients also send files.

I would like to point out that there were no errors in MUC initialization with this code.
```
        ProviderManager.addExtensionProvider("x", "http://jabber.org/protocol/muc#user",
                new org.jivesoftware.smackx.muc.provider.MUCUserProvider());
        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/muc#admin",
                new org.jivesoftware.smackx.muc.provider.MUCAdminProvider());
        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/muc#owner",
                new org.jivesoftware.smackx.muc.provider.MUCOwnerProvider());

```

Appeared with these changes

```
        ProviderManager.addExtensionProvider("x", "http://jabber.org/protocol/muc#user", new MUCUserProvider());
        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/muc#admin", new MUCAdminProvider());
        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/muc#owner", new MUCOwnerProvider());
```

Log: using publishXMPP, publishGroupXMPP, publishXMPPImageByHTTP
```
2026-04-09 11:34:35.537 [INFO ] [org.openhab.core.Activator          ] - Starting openHAB 5.2.0.M2 (Milestone Build)
2026-04-09 11:34:35.883 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Time zone set to 'Etc/GMT+3'.
2026-04-09 11:34:35.884 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Locale set to 'en_US'.
2026-04-09 11:34:38.583 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'lab_x13.items'
2026-04-09 11:34:38.710 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient.items'
2026-04-09 11:34:39.004 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient.things'
2026-04-09 11:34:41.403 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient_sendgroup.script'
2026-04-09 11:34:41.732 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient_senimage.script'
2026-04-09 11:34:41.806 [INFO ] [.core.model.lsp.internal.ModelServer] - Started Language Server Protocol (LSP) service on port 5007
2026-04-09 11:34:41.891 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient.rules'
2026-04-09 11:34:43.311 [WARN ] [.discovery.sddp.SddpDiscoveryService] - listenActiveScanUnicast() error 'Invalid argument: setsockopt'
2026-04-09 11:34:44.087 [INFO ] [t.internal.handler.XMPPClientHandler] - XMPP added channel xmppclient:xmppBridge:homebot:messageEvent payload null
2026-04-09 11:34:51.142 [INFO ] [e.automation.internal.RuleEngineImpl] - Rule engine started.
2026-04-09 11:35:53.360 [INFO ] [t.internal.handler.XMPPClientHandler] - XMPP added channel xmppclient:xmppBridge:homebot:messageEvent payload null
2026-04-09 11:43:39.729 [ERROR] [e.smackx.formtypes.FormFieldRegistry] - Could not load org.igniterealtime.smack/xdata/form-registry/muc/muc-register.xml
java.io.IOException: Unable to get 'org.igniterealtime.smack/xdata/form-registry/muc/muc-register.xml' from classpath. Tried ClassLoaders:[org.eclipse.osgi.internal.loader.EquinoxClassLoader@54004d4d[org.igniterealtime.smack.smack-core:4.5.0.rc1(id=282)], org.eclipse.osgi.internal.framework.ContextFinder@2e6ba49a]
	at org.jivesoftware.smack.util.FileUtils.getStreamForClasspathFile(FileUtils.java:74)
	at org.jivesoftware.smack.util.FileUtils.getStreamForClasspathFile(FileUtils.java:58)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.parseRegistryPath(FormFieldRegistry.java:122)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.loadFormFieldRegistryEntries(FormFieldRegistry.java:154)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.<clinit>(FormFieldRegistry.java:55)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parseField(DataFormProvider.java:258)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parseField(DataFormProvider.java:147)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parse(DataFormProvider.java:95)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parse(DataFormProvider.java:64)
	at org.jivesoftware.smack.provider.Provider.lambda$parse$0(Provider.java:56)
	at org.jivesoftware.smack.provider.AbstractProvider.wrapExceptions(AbstractProvider.java:105)
	at org.jivesoftware.smack.provider.Provider.parse(Provider.java:56)
	at org.jivesoftware.smack.util.PacketParserUtils.parseExtensionElement(PacketParserUtils.java:829)
	at org.jivesoftware.smack.util.PacketParserUtils.addExtensionElement(PacketParserUtils.java:905)
	at org.jivesoftware.smack.util.PacketParserUtils.addExtensionElement(PacketParserUtils.java:900)
	at org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider.parse(DiscoverInfoProvider.java:74)
	at org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider.parse(DiscoverInfoProvider.java:41)
	at org.jivesoftware.smack.provider.IqProvider.lambda$parse$0(IqProvider.java:107)
	at org.jivesoftware.smack.provider.AbstractProvider.wrapExceptions(AbstractProvider.java:105)
	at org.jivesoftware.smack.provider.IqProvider.parse(IqProvider.java:107)
	at org.jivesoftware.smack.util.PacketParserUtils.parseIQ(PacketParserUtils.java:549)
	at org.jivesoftware.smack.util.PacketParserUtils.parseStanza(PacketParserUtils.java:110)
	at org.jivesoftware.smack.AbstractXMPPConnection.parseAndProcessStanza(AbstractXMPPConnection.java:1474)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection.access$100(XMPPTCPConnection.java:132)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.parsePackets(XMPPTCPConnection.java:987)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader$1.run(XMPPTCPConnection.java:952)
	at java.base/java.lang.Thread.run(Thread.java:1583)
2026-04-09 11:43:39.735 [ERROR] [e.smackx.formtypes.FormFieldRegistry] - Could not load org.igniterealtime.smack/xdata/form-registry/muc/muc-request.xml
java.io.IOException: Unable to get 'org.igniterealtime.smack/xdata/form-registry/muc/muc-request.xml' from classpath. Tried ClassLoaders:[org.eclipse.osgi.internal.loader.EquinoxClassLoader@54004d4d[org.igniterealtime.smack.smack-core:4.5.0.rc1(id=282)], org.eclipse.osgi.internal.framework.ContextFinder@2e6ba49a]
	at org.jivesoftware.smack.util.FileUtils.getStreamForClasspathFile(FileUtils.java:74)
	at org.jivesoftware.smack.util.FileUtils.getStreamForClasspathFile(FileUtils.java:58)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.parseRegistryPath(FormFieldRegistry.java:122)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.loadFormFieldRegistryEntries(FormFieldRegistry.java:154)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.<clinit>(FormFieldRegistry.java:55)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parseField(DataFormProvider.java:258)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parseField(DataFormProvider.java:147)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parse(DataFormProvider.java:95)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parse(DataFormProvider.java:64)
	at org.jivesoftware.smack.provider.Provider.lambda$parse$0(Provider.java:56)
	at org.jivesoftware.smack.provider.AbstractProvider.wrapExceptions(AbstractProvider.java:105)
	at org.jivesoftware.smack.provider.Provider.parse(Provider.java:56)
	at org.jivesoftware.smack.util.PacketParserUtils.parseExtensionElement(PacketParserUtils.java:829)
	at org.jivesoftware.smack.util.PacketParserUtils.addExtensionElement(PacketParserUtils.java:905)
	at org.jivesoftware.smack.util.PacketParserUtils.addExtensionElement(PacketParserUtils.java:900)
	at org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider.parse(DiscoverInfoProvider.java:74)
	at org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider.parse(DiscoverInfoProvider.java:41)
	at org.jivesoftware.smack.provider.IqProvider.lambda$parse$0(IqProvider.java:107)
	at org.jivesoftware.smack.provider.AbstractProvider.wrapExceptions(AbstractProvider.java:105)
	at org.jivesoftware.smack.provider.IqProvider.parse(IqProvider.java:107)
	at org.jivesoftware.smack.util.PacketParserUtils.parseIQ(PacketParserUtils.java:549)
	at org.jivesoftware.smack.util.PacketParserUtils.parseStanza(PacketParserUtils.java:110)
	at org.jivesoftware.smack.AbstractXMPPConnection.parseAndProcessStanza(AbstractXMPPConnection.java:1474)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection.access$100(XMPPTCPConnection.java:132)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.parsePackets(XMPPTCPConnection.java:987)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader$1.run(XMPPTCPConnection.java:952)
	at java.base/java.lang.Thread.run(Thread.java:1583)
2026-04-09 11:43:39.738 [ERROR] [e.smackx.formtypes.FormFieldRegistry] - Could not load org.igniterealtime.smack/xdata/form-registry/muc/muc-roomconfig.xml
java.io.IOException: Unable to get 'org.igniterealtime.smack/xdata/form-registry/muc/muc-roomconfig.xml' from classpath. Tried ClassLoaders:[org.eclipse.osgi.internal.loader.EquinoxClassLoader@54004d4d[org.igniterealtime.smack.smack-core:4.5.0.rc1(id=282)], org.eclipse.osgi.internal.framework.ContextFinder@2e6ba49a]
	at org.jivesoftware.smack.util.FileUtils.getStreamForClasspathFile(FileUtils.java:74)
	at org.jivesoftware.smack.util.FileUtils.getStreamForClasspathFile(FileUtils.java:58)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.parseRegistryPath(FormFieldRegistry.java:122)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.loadFormFieldRegistryEntries(FormFieldRegistry.java:154)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.<clinit>(FormFieldRegistry.java:55)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parseField(DataFormProvider.java:258)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parseField(DataFormProvider.java:147)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parse(DataFormProvider.java:95)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parse(DataFormProvider.java:64)
	at org.jivesoftware.smack.provider.Provider.lambda$parse$0(Provider.java:56)
	at org.jivesoftware.smack.provider.AbstractProvider.wrapExceptions(AbstractProvider.java:105)
	at org.jivesoftware.smack.provider.Provider.parse(Provider.java:56)
	at org.jivesoftware.smack.util.PacketParserUtils.parseExtensionElement(PacketParserUtils.java:829)
	at org.jivesoftware.smack.util.PacketParserUtils.addExtensionElement(PacketParserUtils.java:905)
	at org.jivesoftware.smack.util.PacketParserUtils.addExtensionElement(PacketParserUtils.java:900)
	at org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider.parse(DiscoverInfoProvider.java:74)
	at org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider.parse(DiscoverInfoProvider.java:41)
	at org.jivesoftware.smack.provider.IqProvider.lambda$parse$0(IqProvider.java:107)
	at org.jivesoftware.smack.provider.AbstractProvider.wrapExceptions(AbstractProvider.java:105)
	at org.jivesoftware.smack.provider.IqProvider.parse(IqProvider.java:107)
	at org.jivesoftware.smack.util.PacketParserUtils.parseIQ(PacketParserUtils.java:549)
	at org.jivesoftware.smack.util.PacketParserUtils.parseStanza(PacketParserUtils.java:110)
	at org.jivesoftware.smack.AbstractXMPPConnection.parseAndProcessStanza(AbstractXMPPConnection.java:1474)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection.access$100(XMPPTCPConnection.java:132)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.parsePackets(XMPPTCPConnection.java:987)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader$1.run(XMPPTCPConnection.java:952)
	at java.base/java.lang.Thread.run(Thread.java:1583)
2026-04-09 11:43:39.741 [ERROR] [e.smackx.formtypes.FormFieldRegistry] - Could not load org.igniterealtime.smack/xdata/form-registry/muc/muc-roominfo.xml
java.io.IOException: Unable to get 'org.igniterealtime.smack/xdata/form-registry/muc/muc-roominfo.xml' from classpath. Tried ClassLoaders:[org.eclipse.osgi.internal.loader.EquinoxClassLoader@54004d4d[org.igniterealtime.smack.smack-core:4.5.0.rc1(id=282)], org.eclipse.osgi.internal.framework.ContextFinder@2e6ba49a]
	at org.jivesoftware.smack.util.FileUtils.getStreamForClasspathFile(FileUtils.java:74)
	at org.jivesoftware.smack.util.FileUtils.getStreamForClasspathFile(FileUtils.java:58)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.parseRegistryPath(FormFieldRegistry.java:122)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.loadFormFieldRegistryEntries(FormFieldRegistry.java:154)
	at org.jivesoftware.smackx.formtypes.FormFieldRegistry.<clinit>(FormFieldRegistry.java:55)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parseField(DataFormProvider.java:258)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parseField(DataFormProvider.java:147)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parse(DataFormProvider.java:95)
	at org.jivesoftware.smackx.xdata.provider.DataFormProvider.parse(DataFormProvider.java:64)
	at org.jivesoftware.smack.provider.Provider.lambda$parse$0(Provider.java:56)
	at org.jivesoftware.smack.provider.AbstractProvider.wrapExceptions(AbstractProvider.java:105)
	at org.jivesoftware.smack.provider.Provider.parse(Provider.java:56)
	at org.jivesoftware.smack.util.PacketParserUtils.parseExtensionElement(PacketParserUtils.java:829)
	at org.jivesoftware.smack.util.PacketParserUtils.addExtensionElement(PacketParserUtils.java:905)
	at org.jivesoftware.smack.util.PacketParserUtils.addExtensionElement(PacketParserUtils.java:900)
	at org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider.parse(DiscoverInfoProvider.java:74)
	at org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider.parse(DiscoverInfoProvider.java:41)
	at org.jivesoftware.smack.provider.IqProvider.lambda$parse$0(IqProvider.java:107)
	at org.jivesoftware.smack.provider.AbstractProvider.wrapExceptions(AbstractProvider.java:105)
	at org.jivesoftware.smack.provider.IqProvider.parse(IqProvider.java:107)
	at org.jivesoftware.smack.util.PacketParserUtils.parseIQ(PacketParserUtils.java:549)
	at org.jivesoftware.smack.util.PacketParserUtils.parseStanza(PacketParserUtils.java:110)
	at org.jivesoftware.smack.AbstractXMPPConnection.parseAndProcessStanza(AbstractXMPPConnection.java:1474)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection.access$100(XMPPTCPConnection.java:132)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader.parsePackets(XMPPTCPConnection.java:987)
	at org.jivesoftware.smack.tcp.XMPPTCPConnection$PacketReader$1.run(XMPPTCPConnection.java:952)
	at java.base/java.lang.Thread.run(Thread.java:1583)
2026-04-09 11:43:39.743 [WARN ] [ackx.xdata.provider.DataFormProvider] - The Field 'max-file-size' from FORM_TYPE 'urn:xmpp:http:upload:0' is not registered. Field type is unknown, assuming text-single.
2026-04-09 11:43:39.750 [ERROR] [dule.handler.AnnotationActionHandler] - Could not call method 'public void org.openhab.binding.xmppclient.internal.action.XMPPActions.publishXMPPImageByHTTP(java.lang.String,java.lang.String)' from module type 'xmppclient.publishXMPPImageByHTTP#26aa5ea09027ad2b11b752652e808c6a'.
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:109)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.openhab.core.automation.internal.module.handler.AnnotationActionHandler.execute(AnnotationActionHandler.java:127)
	at org.openhab.core.automation.rest.internal.ThingActionsResource.executeThingAction(ThingActionsResource.java:259)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179)
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:307)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267)
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:304)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:517)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:279)
	at org.ops4j.pax.web.service.spi.servlet.OsgiInitializedServlet.service(OsgiInitializedServlet.java:102)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
	at org.ops4j.pax.web.service.spi.servlet.OsgiFilterChain.doFilter(OsgiFilterChain.java:113)
	at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler.doHandle(PaxWebServletHandler.java:334)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:234)
	at org.ops4j.pax.web.service.jetty.internal.PrioritizedHandlerCollection.handle(PrioritizedHandlerCollection.java:96)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:722)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class org.jivesoftware.smack.packet.UnparsedIQ cannot be cast to class org.jivesoftware.smackx.httpfileupload.element.Slot (org.jivesoftware.smack.packet.UnparsedIQ is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @54004d4d; org.jivesoftware.smackx.httpfileupload.element.Slot is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @27a71274)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.requestSlot(HttpFileUploadManager.java:423)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.requestSlot(HttpFileUploadManager.java:358)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.uploadFile(HttpFileUploadManager.java:266)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.uploadFile(HttpFileUploadManager.java:243)
	at org.openhab.binding.xmppclient.internal.client.XMPPClient.sendImageByHTTP(XMPPClient.java:226)
	at org.openhab.binding.xmppclient.internal.action.XMPPActions.publishXMPPImageByHTTP(XMPPActions.java:100)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 60 more

```

> The same jar works with xmpp.jp (can;t really test the file upload, but hte rest seems to function)

Hi!!!
I've been busy with work for a long time again!

I checked it especially for you.
You can easily test everything on the xmpp.jp server. It works perfectly, unlike binding. I registered two accounts. One logged in with Gahjim (PC), the other with OpenHab 5.2.0M2 with your changes. Photos are not sent. 
You can check for yourself.
Still the same problem, bad runtime version.

Regarding running Openfire locally, you can get a Let's Encrypt certificate if you have an external IP address and configure it in Openfire by forwarding ports on your router.
```
2026-04-26 14:08:33.381 [INFO ] [org.openhab.core.Activator          ] - Starting openHAB 5.2.0.M2 (Milestone Build)
2026-04-26 14:08:33.770 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Time zone set to 'Etc/GMT+3'.
2026-04-26 14:08:33.771 [INFO ] [.core.internal.i18n.I18nProviderImpl] - Locale set to 'en_US'.
2026-04-26 14:08:36.921 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'lab_x13.items'
2026-04-26 14:08:37.057 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient.items'
2026-04-26 14:08:37.446 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient_server1.things'
2026-04-26 14:08:39.949 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient_sendgroup.script'
2026-04-26 14:08:40.338 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient_senimage.script'
2026-04-26 14:08:40.425 [INFO ] [.core.model.lsp.internal.ModelServer] - Started Language Server Protocol (LSP) service on port 5007
2026-04-26 14:08:40.500 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading DSL model 'xmppclient.rules'
2026-04-26 14:08:42.150 [WARN ] [.discovery.sddp.SddpDiscoveryService] - listenActiveScanUnicast() error 'Invalid argument: setsockopt'
2026-04-26 14:08:43.011 [INFO ] [t.internal.handler.XMPPClientHandler] - XMPP added channel xmppclient:xmppBridge:madhomebot:messageEvent payload null
2026-04-26 14:08:49.072 [INFO ] [e.automation.internal.RuleEngineImpl] - Rule engine started.
2026-04-26 14:09:01.594 [WARN ] [.io.rest.auth.internal.TokenResource] - Couldn't find a user with a session matching the provided refresh_token
2026-04-26 14:09:01.596 [WARN ] [.io.rest.auth.internal.TokenResource] - Token issuing failed: invalid_grant
2026-04-26 14:16:16.322 [ERROR] [dule.handler.AnnotationActionHandler] - Could not call method 'public void org.openhab.binding.xmppclient.internal.action.XMPPActions.publishXMPPImageByHTTP(java.lang.String,java.lang.String)' from module type 'xmppclient.publishXMPPImageByHTTP#26aa5ea09027ad2b11b752652e808c6a'.
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:109)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.openhab.core.automation.internal.module.handler.AnnotationActionHandler.execute(AnnotationActionHandler.java:127)
	at org.openhab.core.automation.rest.internal.ThingActionsResource.executeThingAction(ThingActionsResource.java:259)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179)
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:307)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267)
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:304)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:517)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:279)
	at org.ops4j.pax.web.service.spi.servlet.OsgiInitializedServlet.service(OsgiInitializedServlet.java:102)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
	at org.ops4j.pax.web.service.spi.servlet.OsgiFilterChain.doFilter(OsgiFilterChain.java:113)
	at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler.doHandle(PaxWebServletHandler.java:334)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:234)
	at org.ops4j.pax.web.service.jetty.internal.PrioritizedHandlerCollection.handle(PrioritizedHandlerCollection.java:96)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:722)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class org.jivesoftware.smack.packet.UnparsedIQ cannot be cast to class org.jivesoftware.smackx.httpfileupload.element.Slot (org.jivesoftware.smack.packet.UnparsedIQ is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @9107e3a; org.jivesoftware.smackx.httpfileupload.element.Slot is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @7003e9ca)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.requestSlot(HttpFileUploadManager.java:423)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.requestSlot(HttpFileUploadManager.java:358)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.uploadFile(HttpFileUploadManager.java:266)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.uploadFile(HttpFileUploadManager.java:243)
	at org.openhab.binding.xmppclient.internal.client.XMPPClient.sendImageByHTTP(XMPPClient.java:226)
	at org.openhab.binding.xmppclient.internal.action.XMPPActions.publishXMPPImageByHTTP(XMPPActions.java:100)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 60 more
2026-04-26 14:16:38.478 [ERROR] [dule.handler.AnnotationActionHandler] - Could not call method 'public void org.openhab.binding.xmppclient.internal.action.XMPPActions.publishXMPPImageByHTTP(java.lang.String,java.lang.String)' from module type 'xmppclient.publishXMPPImageByHTTP#26aa5ea09027ad2b11b752652e808c6a'.
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:109)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.openhab.core.automation.internal.module.handler.AnnotationActionHandler.execute(AnnotationActionHandler.java:127)
	at org.openhab.core.automation.rest.internal.ThingActionsResource.executeThingAction(ThingActionsResource.java:259)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:179)
	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:201)
	at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:104)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)
	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:96)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:307)
	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:267)
	at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:234)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:208)
	at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:160)
	at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:225)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:304)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:517)
	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:279)
	at org.ops4j.pax.web.service.spi.servlet.OsgiInitializedServlet.service(OsgiInitializedServlet.java:102)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
	at org.ops4j.pax.web.service.spi.servlet.OsgiFilterChain.doFilter(OsgiFilterChain.java:113)
	at org.ops4j.pax.web.service.jetty.internal.PaxWebServletHandler.doHandle(PaxWebServletHandler.java:334)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:234)
	at org.ops4j.pax.web.service.jetty.internal.PrioritizedHandlerCollection.handle(PrioritizedHandlerCollection.java:96)
	at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:722)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class org.jivesoftware.smack.packet.UnparsedIQ cannot be cast to class org.jivesoftware.smackx.httpfileupload.element.Slot (org.jivesoftware.smack.packet.UnparsedIQ is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @9107e3a; org.jivesoftware.smackx.httpfileupload.element.Slot is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @7003e9ca)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.requestSlot(HttpFileUploadManager.java:423)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.requestSlot(HttpFileUploadManager.java:358)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.uploadFile(HttpFileUploadManager.java:266)
	at org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager.uploadFile(HttpFileUploadManager.java:243)
	at org.openhab.binding.xmppclient.internal.client.XMPPClient.sendImageByHTTP(XMPPClient.java:226)
	at org.openhab.binding.xmppclient.internal.action.XMPPActions.publishXMPPImageByHTTP(XMPPActions.java:100)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 60 more

```

--- a/bundles/org.openhab.binding.xmppclient/pom.xml
+++ b/bundles/org.openhab.binding.xmppclient/pom.xml
@@ -70,6 +70,19 @@
       <scope>provided</scope>
     </dependency>
     <dependency>
+      <groupId>org.igniterealtime.smack</groupId>
+      <artifactId>smack-xmlparser</artifactId>
+      <version>${smack.version}</version>
+      <scope>provided</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.igniterealtime.smack</groupId>
+      <artifactId>smack-xmlparser-xpp3</artifactId>
+      <version>${smack.version}</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
       <groupId>org.minidns</groupId>
       <artifactId>minidns-core</artifactId>
       <version>1.1.1</version>

--- a/bundles/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/XMPPClientHandlerFactory.java
+++ b/bundles/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/XMPPClientHandlerFactory.java
@@ -26,8 +26,13 @@ import org.jivesoftware.smack.util.stringencoder.Base64;
 import org.jivesoftware.smack.util.stringencoder.Base64UrlSafeEncoder;
 import org.jivesoftware.smack.util.stringencoder.java7.Java7Base64Encoder;
 import org.jivesoftware.smack.util.stringencoder.java7.Java7Base64UrlSafeEncoder;
+import org.jivesoftware.smack.xml.SmackXmlParser;
+import org.jivesoftware.smack.xml.xpp3.Xpp3XmlPullParserFactory;
 import org.jivesoftware.smackx.disco.provider.DiscoverInfoProvider;
 import org.jivesoftware.smackx.disco.provider.DiscoverItemsProvider;
+import org.jivesoftware.smackx.muc.provider.MUCAdminProvider;
+import org.jivesoftware.smackx.muc.provider.MUCOwnerProvider;
+import org.jivesoftware.smackx.muc.provider.MUCUserProvider;
 import org.jivesoftware.smackx.xdata.provider.DataFormProvider;
 import org.openhab.binding.xmppclient.internal.handler.XMPPClientHandler;
 import org.openhab.core.thing.Bridge;
@@ -70,13 +75,22 @@ public class XMPPClientHandlerFactory extends BaseThingHandlerFactory {
     @Activate
     protected void activate(ComponentContext componentContext) {
         super.activate(componentContext);
+        SmackXmlParser.setXmlPullParserFactory(new Xpp3XmlPullParserFactory());
+
         Roster.setRosterLoadedAtLoginDefault(false);
-        DNSUtil.setDNSResolver(JavaxResolver.getInstance());
+        var dnsResolver = JavaxResolver.getInstance();
+        if (dnsResolver != null) {
+            DNSUtil.setDNSResolver(dnsResolver);
+        }
         SmackConfiguration.setDefaultHostnameVerifier(new XmppHostnameVerifier());
         Base64.setEncoder(Java7Base64Encoder.getInstance());
         Base64UrlSafeEncoder.setEncoder(Java7Base64UrlSafeEncoder.getInstance());
         ProviderManager.addIQProvider("query", "http://jabber.org/protocol/disco#info", new DiscoverInfoProvider());
         ProviderManager.addIQProvider("query", "http://jabber.org/protocol/disco#items", new DiscoverItemsProvider());
         ProviderManager.addExtensionProvider("x", "jabber:x:data", new DataFormProvider());
+
+        ProviderManager.addExtensionProvider("x", "http://jabber.org/protocol/muc#user", new MUCUserProvider());
+        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/muc#admin", new MUCAdminProvider());
+        ProviderManager.addIQProvider("query", "http://jabber.org/protocol/muc#owner", new MUCOwnerProvider());
     }
 }

--- a/bundles/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/client/XMPPClient.java
+++ b/bundles/org.openhab.binding.xmppclient/src/main/java/org/openhab/binding/xmppclient/internal/client/XMPPClient.java
@@ -188,24 +188,60 @@ public class XMPPClient implements IncomingChatMessageListener, ConnectionListen
     }
 
     public void sendImageByHTTP(String to, String filename) {
-        if (connection == null) {
-            logger.warn("XMPP connection is null");
+        if (connection == null || !connection.isAuthenticated()) {
+            logger.warn("XMPP connection is null or not authenticated");
             return;
         }
-        HttpFileUploadManager httpFileUploadManagerLocal = httpFileUploadManager;
-        if (httpFileUploadManagerLocal == null) {
-            httpFileUploadManagerLocal = httpFileUploadManager = HttpFileUploadManager.getInstanceFor(connection);
-            if (httpFileUploadManagerLocal == null) {
-                logger.warn("XMPP httpFileUploadManager is null");
+
+        File file = new File(filename);
+        if (!file.isFile() || !file.canRead()) {
+            logger.warn("XMPP image file does not exist or is not readable: {}", filename);
+            return;
+        }
+
+        HttpFileUploadManager uploadManager = httpFileUploadManager;
+        if (uploadManager == null) {
+            httpFileUploadManager = uploadManager = HttpFileUploadManager.getInstanceFor(connection);
+        }
+
+        if (uploadManager == null) {
+            logger.warn("XMPP HttpFileUploadManager is null");
+            return;
+        }
+
+        try {
+            // Force discovery instead of assuming it already happened
+            if (!uploadManager.isUploadServiceDiscovered()) {
+                boolean discovered = uploadManager.discoverUploadService();
+                logger.debug("XMPP HTTP upload discovered: {}", discovered);
+            }
+
+            if (!uploadManager.isUploadServiceDiscovered()) {
+                logger.warn("XMPP server does not provide XEP-0363 HTTP File Upload");
                 return;
             }
-        }
-        try {
-            URL u = httpFileUploadManagerLocal.uploadFile(new File(filename));
-            // Use Stanza oob
-            this.sendMessage(to, u.toString());
-        } catch (XMPPException.XMPPErrorException | SmackException | InterruptedException | IOException e) {
-            logger.warn("XMPP HTTP image sending error", e);
+
+            logger.debug("XMPP HTTP upload service: {}", uploadManager.getDefaultUploadService());
+
+            URL url = uploadManager.uploadFile(file);
+
+            // Simple fallback: just send the URL as a message.
+            // If you want true out-of-band metadata, build an OOB extension stanza instead.
+            sendMessage(to, url.toString());
+
+        } catch (XMPPException.XMPPErrorException e) {
+            logger.warn("XMPP HTTP upload rejected by server: {}", e.getMessage(), e);
+        } catch (SmackException.NoResponseException e) {
+            logger.warn("XMPP HTTP upload timed out while discovering service or requesting slot", e);
+        } catch (SmackException.NotConnectedException e) {
+            logger.warn("XMPP HTTP upload failed because connection is not active", e);
+        } catch (SmackException e) {
+            logger.warn("XMPP HTTP upload failed: {}", e.getMessage(), e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            logger.warn("XMPP HTTP upload interrupted", e);
+        } catch (IOException e) {
+            logger.warn("XMPP HTTP upload I/O error", e);
         }
     }
 

